### ISSUE TITLE
[meta] Consider allowing users to ignore SSL errors via prompt

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/focus-android#1616]
_Added by @bbinto_

**Why/User Benefit**
- Provide users with an option to still access the page even with an SSL error

**What**
- Either let them open the page *not* with Focus but another browser or with Focus but with educating the users about their risks doing so

**Acceptance Criteria (how do I know when I’m done?)**
- User can access web page after SSL error was shown
- Probe to track how often users end up on pages like that

----
Clone of #1563:

When there's an SSL error, we currently show:

![screenshot_1508977136](https://user-images.githubusercontent.com/759372/32029328-a4843150-b9a8-11e7-9db0-7e38effb33ec.png)

Desktop browsers often let the user ignore SSL errors, however:

<img width="1105" alt="screen shot 2017-10-25 at 17 02 35" src="https://user-images.githubusercontent.com/759372/32029339-b1d3ae80-b9a8-11e7-9119-762132816a7e.png">

We should consider doing something similar.

One option would be to encourage users to go to a different browser to ignore SSL errors so Focus can try to protect the user. Note: as far as I can tell, we don't cache intermediate CAs, which means we're more likely to get SSL errors than other browsers (investigation in #1338).

Here's the same issue on iOS: https://github.com/mozilla-mobile/focus-ios/issues/238

**Comment by brampitoyo** [comment]
My suggestion is to implement a similar design that’s been responsified, but have “Open In…” functionality in place of “Add Exception…”. This allows users to open the page in a browser that can ignore SSL errors.

<img width="784" alt="screen shot 2018-03-13 at 5 32 17 pm 2x" src="https://user-images.githubusercontent.com/51007/37322826-cefac0f8-26e4-11e8-86d3-211d74077e55.png">

We can simplify the page even further by showing “Open In…” on its initial state, at the cost of skipping the error details entirely:

<img src="https://user-images.githubusercontent.com/51007/37322937-7e6e179c-26e5-11e8-88db-688380ba9809.png" width="360">

**Comment by bbinto** [comment]
Are we still blocked by GeckoView, @mcomella?

**Comment by brampitoyo** [comment]
Just mentioning the fact that this issue is ready for Engineering breakdown.

I’ll provide design specs very soon!

**Comment by bbinto** [comment]
Are we still blocked on GeckoView, @pocmo?

**Comment by pocmo** [comment]
I expect GeckoView to open an "about:neterror" page with additional parameters about the error in the URL. As part of the breakdown an engineer could test this and see what values GeckoView provides us and then we may have to update the mock accordingly. There are potentially a gazillion possible SSL errors so we may not be able to show detailed information for every one of them. I also assume that GeckoView won't report anything back to Mozilla by default so we may not want to show that checkbox.

@brampitoyo Assuming that the SSL error data GeckoView provides will be pretty limited I think your simplified page is probably the way to go. I wonder what "Go Back" will actually do though?

**Comment by brampitoyo** [comment]
> There are potentially a gazillion possible SSL errors so we may not be able to show detailed information for every one of them.

In this case, we may not have to show the advanced page at all – making the design even simpler.

Here’s the updated design that shows some of what we’ve been thinking:

* No “Advanced” section – it’s been replaced by “Open In”
* No “Report” checkbox
* And an idea: what if “Open In” is replaced with **“Accept the Risk and Continue”**? It will only work once, though. You won’t be able to add a permanent exception.

<img width="796" alt="screen shot 2018-06-21 at 4 44 19 pm" src="https://user-images.githubusercontent.com/51007/41698475-afe73d04-7572-11e8-8c29-9b4fd9e6b24d.png">

**Comment by pocmo** [comment]
> And an idea: what if “Open In” is replaced with “Accept the Risk and Continue”?

That's something we would have to bring up with the GeckoView team: I'm not sure how/if we are able to do the override currently.

**Comment by brampitoyo** [comment]
@pocmo There are good arguments both for and against having the “Accept the Risk” feature.

Against “Accept the Risk”: opening risky links doesn’t sound like something that a privacy-focused browser should encourage.

Pro “Accept the Risk”: there’s no safer way to open risky links than by using Firefox Focus, because we can block malicious elements by default, and the user can back out in one step – just tap Erase.

For now, it would be good to support “Open In…”; In the future, we should ask the GeckoView team if they plan to support ignoring SSL errors.

**Comment by pocmo** [comment]
> For now, it would be good to support “Open In…”; In the future, we should ask the GeckoView team if they plan to support ignoring SSL errors.

That was the result of the meeting with the GV team too. We may be able to implement the version with "Open in..." now without new APIs and later can try to build the "risk override" version once there is a more evolved "custom error page API".

For option “Open In…” we would:
 * Intercept about:certerror loads so that we can inject our own error page
 * Create a new error page for the new design above (Ideally we move this into the browser-errorpages component)
 * Right now there's no way how we can communicate that "open in" was clicked via a GeckoView API. However we could try to make this load a custom protocol (like other app links) and then implement the open in behavior based on that. (Maybe we can just trigger our normal "You are leaving Focus" popup?)

Edge cases:
* What if there is no other browser installed? Very unlikely but technically possible. (Or maybe there is another browser but we couldn't find it..)
* What if there is only one other browser installed - would we still show our "open with" sheet or would we open the other browser immediately.

**Comment by brampitoyo** [comment]
> What if there is no other browser installed? Very unlikely but technically possible.

This is a bit problematic. I was hoping that every manufacturer would ship with a general purpose browser preinstalled, and that Focus will be the one either set as a secondary, or one downloaded by the user later.

Can we somehow take advantage of “plain webview” to offer a browser-within-a-browser experience? This feels overthinking and overcomplicating things – but it’s an edge case, after all!


> What if there is only one other browser installed - would we still show our "open with" sheet or would we open the other browser immediately.

It would be ideal if we can modify the button label to say “Open in [browser_name]”. Then it’s accurate.

Otherwise, the expectation when tapping a button labelled “Open in…” is to see some sort of a menu or a dialogue box, not to instantly be taken to another app.

TL;DR – If we can show the browser name, let’s open it immediately. If we can’t show the name, let’s keep showing the existing menu.

**Comment by brampitoyo** [comment]
Spec
<img src="https://user-images.githubusercontent.com/51007/42199773-7512565c-7ee4-11e8-8931-94b3df454a4c.png" width="360">
